spec/app/domain/authentication/authn_k8s/execute_command_in_container_spec.rb

```diff
@@ -0,0 +1,325 @@
+    end
+  end
+
+  let(:message) do
+    double('Message').tap do |message|
+      allow(message).to receive(:type).and_return(:binary)
+      allow(message).to receive(:data).and_return(
+        # we need to start the message with a zero byte to indicate stdin
+        "#{[0].pack('c*')}some message"
+      )
+    end
+  end
+
+  let(:kube_client_headers) do
+    double('KubeClientHeaders').tap do |kube_client_headers|
+      allow(kube_client_headers).to receive(:clone)
+        .and_return([])
+    end
+  end
+
+  let(:kube_client_api_endpoint) do
+    double('kube_client_api_endpoint').tap do |kube_client_api_endpoint|
+      allow(kube_client_api_endpoint).to receive(:host)
+        .and_return("host")
+      allow(kube_client_api_endpoint).to receive(:port)
+        .and_return("port")
+    end
+  end
+
+  let(:kube_client) do
+    double('KubeClient').tap do |kube_client|
+      allow(kube_client).to receive(:headers)
+        .and_return(kube_client_headers)
+      allow(kube_client).to receive(:api_endpoint)
+        .and_return(kube_client_api_endpoint)
+    end
+  end
+
+  let(:k8s_object_lookup) do
+    double('K8sObjectLookup').tap do |k8s_object_lookup|
+      allow(k8s_object_lookup).to receive(:kube_client)
+        .and_return(kube_client)
+    end
+  end
+
+  let(:pod_namespace) { "PodNamespace" }
+  let(:pod_name) { "PodName" }
+  let(:container) { "Container" }
+  let(:cmds) { %w(command1 command2) }
+  let(:body) { "Body" }
+
+  # This is used in test code to verify that all the client listeners are ready.
+  # We wait for this timeout while verifying this
+  let(:ready_listeners_timeout) { 5 }
+
+  def subject_in_thread(ws_client:, timeout:, body:, stdin:)
+    # This is ugly but needed because references to "let"-defined variables
+    # inside of a closure (or at least a closure in a Thread) don't work.
+    # References to ordinary local variables work fine.
+    x_timeout           = timeout
+    x_ws_client         = ws_client
+    x_k8s_object_lookup = k8s_object_lookup
+    x_pod_namespace     = pod_namespace
+    x_pod_name          = pod_name
+    x_container         = container
+    x_cmds              = cmds
+    x_body              = body
+    x_stdin             = stdin
+
+    Thread.new do
+      Thread.current[:output] =
+        ::Authentication::AuthnK8s::ExecuteCommandInContainer.new(
+          timeout:          x_timeout,
+          websocket_client: x_ws_client,
+        ).call(
+          k8s_object_lookup: x_k8s_object_lookup,
+          pod_namespace:     x_pod_namespace,
+          pod_name:          x_pod_name,
+          container:         x_container,
+          cmds:              x_cmds,
+          body:              x_body,
+          stdin:             x_stdin
+        )
+    end.tap do
+      # Verify all 4 listeners are ready before resuming the UT
+      Timeout.timeout(ready_listeners_timeout) {
+        4.times do
+          ws_client.ready_listeners_queue.pop
+        end
+      }
```

Comment:
This code should be moved into the mock itself under a method like `wait_until_ready` with a comment like:

```
We need to wait until the implementation code has set all of the event
handlers.  This allows us to block until they have all been called.
```
Also, more importantly, since the logic is now "we want to wait until _every handler has been set_", each event handler should add a string event with its own event name to the queue.  Then the "wait" method would pop until all 4 uniq events have been received.  The way the code is currently, it's coupled to the existing implementation.  If someone changed the code to, say, call a single event handler 4 times, the code would be convinced everything was set.